Add new menu type field matrix

khanning opened the pull request:
### Proposed Changes

Create a new field type for 5x5 LED matrix displays.
![image](https://user-images.githubusercontent.com/1169507/42356667-bb6e0e20-80a1-11e8-937d-180dff712484.png)

### Reason for Changes

For interacting with hardware devices that have programmable LED matrix displays (e.g. micro:bit).

### Test Coverage

#### OS X (High Sierra 10.13.4)
  - Chrome (67.0.3396.99)
  - Firefox (52.9.0)
  - Safari (11.1)

#### Android (8.1.0)
  - Chrome (67.0.3396.87)

khanning committed:
Add new menu type field matrix

khanning committed:
Add sample matrix block to the default toolbox

khanning committed:
Fix source block colour scheme bug

khanning commented:
Thanks to @rachel-fenichel the colour scheme issue is fixed! The PR can now be reviewed.

rachel-fenichel commented on core/field_matrix.js:
```diff
@@ -0,0 +1,456 @@
+
+  this.getSourceColours_();
+
+  if (!this.ledThumbs_) {
```
General comment:
Any properties that the field has should be explicitly declared and annotated, often in the constructor, unless they're declared in the superclass.

This isn't done in most of the existing fields because they got created before I understood how helpful these annotations are.

Benefits:
- When debugging down the line, you won't need to search for the first place that a property is used to know what it is
- Automated type checking can catch bugs (if it's being used)
- The next person editing this code won't come find where you live to understand it
- You catch places where you're mixing types in an unsafe way
- You can explicitly set the default value, rather than checking for undefined

Downsides:
- Constructors become gigantic, and most of it is in comments
- Takes time
- ???

Blockly's [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide#jsdoc) describes what annotations should look like

ericrosenbaum commented on blocks_vertical/default_toolbox.js:
```diff
@@ -541,5 +541,12 @@ Blockly.Blocks.defaultToolbox = '<xml id="toolbox-categories" style="display: no
+    '<block type="extension_microbit_display" id="extension_microbit_display">'+
+      '<value name="MATRIX">'+
+        '<shadow type="matrix">'+
+          '<field name="MATRIX">4540074</field>'+
```
♥

khanning committed:
Merge remote-tracking branch 'upstream/develop' into menu-field-matrix

khanning committed:
Refine field matrix menu

  - Change input type from 25 bit integer to 25 char string.
  - Fix pointer flickering while hovering over field.
  - Include node pad in the matrix editor hit zone.
  - Change color scheme to "pen" to match new extension colors.
  - Define all properties in the constructor.
  - Rename a few functions for clarity.
  - Add additional comments

khanning committed:
Remove duplicate block entry

ericrosenbaum reviewed (approved):
This looks ready to go!
